src/main/java/org/reaktivity/nukleus/mqtt/internal/stream/MqttServerFactory.java

```diff
@@ -265,12 +300,375 @@ private RouteFW resolveTarget(
+                                        .build();
+
+        receiver.accept(signal.typeId(), signal.buffer(), signal.offset(), signal.sizeof());
+    }
+
+    private int decodePacketType(
+        MqttServer server,
+        final long traceId,
+        final long authorization,
+        final long budgetId,
+        final DirectBuffer buffer,
+        final int offset,
+        final int limit)
+    {
+        final MqttPacketFixedHeaderFW packet = mqttPacketFixedHeaderRO.tryWrap(buffer, offset, limit);
+
+        if (packet != null)
+        {
+            final int length = packet.remainingLength();
+            final MqttPacketType packetType = MqttPacketType.valueOf(packet.typeAndFlags() >> 4);
+            final MqttServerDecoder decoder = decodersByPacketType.getOrDefault(packetType, decodeIgnoreOne);
+
+            if (limit - packet.limit() >= length)
+            {
+                server.decoder = decoder;
+            }
+        }
+
+        return offset;
+    }
+
+    private int decodeConnect(
+        MqttServer server,
+        final long traceId,
+        final long authorization,
+        final long budgetId,
+        final DirectBuffer buffer,
+        final int offset,
+        final int limit)
+    {
+        MqttConnectFW mqttConnect = mqttConnectRO.tryWrap(buffer, offset, limit);
+
+        int progress = limit;
+        int reasonCode = 0x00;
+        if (mqttConnect == null || (mqttConnect.flags() & 0b0000_0001) != 0b0000_0000)
+        {
+            reasonCode = 0x81; // malformed packet
+        }
+        else if (!"MQTT".equals(mqttConnect.protocolName().asString()) || mqttConnect.protocolVersion() != 5)
+        {
+            reasonCode = 0x84; // unsupported protocol version
+        }
+        else if (server.connected)
+        {
+            reasonCode = 0x82; // protocol error
+        }
+
+        if (reasonCode == 0)
+        {
+            server.onDecodeConnect(authorization, reasonCode, mqttConnect);
+            server.decoder = decodePacketType;
+            progress = mqttConnect.limit();
+        }
+        else
+        {
+            server.onDecodeConnect(authorization, reasonCode, mqttConnect);
+            server.doNetworkEnd(traceId);
+        }
+
+        return progress;
+    }
+
+    private int decodePublish(
+        MqttServer server,
+        final long traceId,
+        final long authorization,
+        final long budgetId,
+        final DirectBuffer buffer,
+        final int offset,
+        final int limit)
+    {
+        final MqttPublishFW publish = mqttPublishRO.tryWrap(buffer, offset, limit);
+        int progress = limit;
+        if (limit == 0 || publish == null)
+        {
+            server.doEncodeDisconnect(authorization, 0x82);
+            server.doNetworkEnd(traceId);
+        }
+        else
+        {
+            server.onDecodePublish(authorization, publish);
+            progress = publish.limit();
+        }
+
+        server.decoder = decodePacketType;
```

Comment:
Advance decoder in successful `else` clause only, switch to `decodeIgnoreAll` in failure `if` case.